In [18]:
!pip install tensorflow

In [19]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt

# print Tensorflow version
print(tf.__version__)

2.15.0


In [20]:
from tensorflow.keras import datasets
from tensorflow.keras import layers
from tensorflow.keras import models


## Import Cats Vs Dogs

In [21]:
import os

_URL = 'https://storage.googleapis.com/mledu-datasets/cats_and_dogs_filtered.zip'
path_to_zip = tf.keras.utils.get_file('cats_and_dogs.zip', origin=_URL, extract=True)
PATH = os.path.join(os.path.dirname(path_to_zip), 'cats_and_dogs_filtered')

train_dir = os.path.join(PATH, 'train')
test_dir = os.path.join(PATH, 'validation')

IMG_SIZE = (128, 128)

# training data
train_dataset = tf.keras.utils.image_dataset_from_directory(train_dir,
                                                            shuffle=True,
                                                            batch_size=None,
                                                            image_size=IMG_SIZE)
train_images, train_labels = [], []
for train_data in train_dataset:
  train_images.append(train_data[0])
  train_labels.append(train_data[1])
train_images = np.asarray(train_images).astype(np.uint8)
train_labels = np.asarray(train_labels)

# test data
test_dataset = tf.keras.utils.image_dataset_from_directory(test_dir,
                                                           shuffle=True,
                                                           batch_size=None,
                                                           image_size=IMG_SIZE)
test_images, test_labels = [], []
for test_data in test_dataset:
  test_images.append(test_data[0])
  test_labels.append(test_data[1])
test_images = np.asarray(test_images).astype(np.uint8)
test_labels = np.asarray(test_labels)

class_names = ['Cat', 'Dog']

Found 2000 files belonging to 2 classes.
Found 1000 files belonging to 2 classes.


# Normalization to [0, 1]

In [22]:
train_images = train_images / 255
test_images = test_images / 255

***Question 3***

In [23]:
def creat_model(input_shape, learning_rate):
  model = tf.keras.Sequential([
      tf.keras.layers.Flatten(input_shape=input_shape),
      tf.keras.layers.Dense(256, activation = 'relu'),
      tf.keras.layers.Dense(128, activation = 'relu'),
      tf.keras.layers.Dense(2, activation = 'softmax')

  ])

In [24]:
def train_model(model, train_images, train_labels, epochs, validation_split):
  eval = model.fit(train_images, train_labels, epochs = epochs, validation_split = validation_split, verbose = 2)
  return eval

In [25]:
import numpy as np
from tensorflow.keras.optimizers import Adam
from sklearn.model_selection import train_test_split


In [27]:
model_300 = models.Sequential([
    layers.Flatten(input_shape = (128, 128, 3)),
    layers.Dense(256, activation = 'relu'),
    layers.Dense(128, activation = 'relu'),
    layers.Dense(2, activation = 'softmax')
])

opt_300 = Adam(learning_rate = 300)
model_300.compile(optimizer = opt_300, loss = 'sparse_categorical_crossentropy', metrics = ['accuracy'])
model_300_valid = model_300.fit(train_images, train_labels, validation_split = 0.15,epochs = 50)



Epoch 1/50
54/54 [==============================] - 2s 21ms/step - loss: 161485233848320.0000 - accuracy: 0.4947 - val_loss: 3023.4578 - val_accuracy: 0.4867
Epoch 2/50
54/54 [==============================] - 1s 16ms/step - loss: 866.3412 - accuracy: 0.5012 - val_loss: 47.3857 - val_accuracy: 0.4867
Epoch 3/50
54/54 [==============================] - 1s 10ms/step - loss: 50.2572 - accuracy: 0.4906 - val_loss: 9.6496 - val_accuracy: 0.5133
Epoch 4/50
54/54 [==============================] - 1s 10ms/step - loss: 64.2705 - accuracy: 0.4941 - val_loss: 13.4630 - val_accuracy: 0.4867
Epoch 5/50
54/54 [==============================] - 1s 9ms/step - loss: 35.8585 - accuracy: 0.5024 - val_loss: 6.5464 - val_accuracy: 0.4867
Epoch 6/50
54/54 [==============================] - 1s 9ms/step - loss: 36.2317 - accuracy: 0.4941 - val_loss: 11.0594 - val_accuracy: 0.5133
Epoch 7/50
54/54 [==============================] - 1s 9ms/step - loss: 43.6886 - accuracy: 0.5118 - val_loss: 61.5082 - val_accur

For learning rate 300, in the first epoch, it suppossed to be " loss is the maximum and accuracy is the minimum. Loss gradually decrease and accuracy gradually increase from the first epoch to 50th epoch". But, here they have not maintained the rules that need to be. Some times loss decrease then again increases and accuracy increases and again decreases.

In [28]:
model_0 = models.Sequential([
    layers.Flatten (input_shape = (128, 128, 3)),
    layers.Dense(256, activation = 'relu'),
    layers.Dense(128, activation = 'relu'),
    layers.Dense(2, activation = 'softmax')
])

opt_0 = Adam(learning_rate = 0)
model_0.compile(optimizer = opt_0, loss = 'sparse_categorical_crossentropy', metrics = ['accuracy'])
model_0_valid = model_0.fit(train_images, train_labels, validation_split = 0.15, epochs = 50)

Epoch 1/50
54/54 [==============================] - 2s 15ms/step - loss: 0.7374 - accuracy: 0.5100 - val_loss: 0.7456 - val_accuracy: 0.4967
Epoch 2/50
54/54 [==============================] - 1s 10ms/step - loss: 0.7374 - accuracy: 0.5100 - val_loss: 0.7456 - val_accuracy: 0.4967
Epoch 3/50
54/54 [==============================] - 1s 10ms/step - loss: 0.7374 - accuracy: 0.5100 - val_loss: 0.7456 - val_accuracy: 0.4967
Epoch 4/50
54/54 [==============================] - 1s 10ms/step - loss: 0.7374 - accuracy: 0.5100 - val_loss: 0.7456 - val_accuracy: 0.4967
Epoch 5/50
54/54 [==============================] - 1s 9ms/step - loss: 0.7374 - accuracy: 0.5100 - val_loss: 0.7456 - val_accuracy: 0.4967
Epoch 6/50
54/54 [==============================] - 1s 10ms/step - loss: 0.7374 - accuracy: 0.5100 - val_loss: 0.7456 - val_accuracy: 0.4967
Epoch 7/50
54/54 [==============================] - 1s 9ms/step - loss: 0.7374 - accuracy: 0.5100 - val_loss: 0.7456 - val_accuracy: 0.4967
Epoch 8/50
54/5

For learning rate 0, loss and accuracy are the same from epoch 1 to epoch 50.

In [29]:
model_00001 = models.Sequential([
    layers.Flatten(input_shape = (128, 128, 3)),
    layers.Dense(256, activation = 'relu'),
    layers.Dense(128, activation = 'relu'),
    layers.Dense(2, activation = 'softmax')

])

opt_00001 = Adam(learning_rate = 0.00001)
model_00001.compile(optimizer = opt_00001, loss = 'sparse_categorical_crossentropy', metrics = ['accuracy']),
model_00001_valid = model_00001.fit(train_images, train_labels, validation_split = 0.15, epochs = 50)

Epoch 1/50
54/54 [==============================] - 2s 16ms/step - loss: 0.7247 - accuracy: 0.5400 - val_loss: 0.6930 - val_accuracy: 0.5633
Epoch 2/50
54/54 [==============================] - 1s 10ms/step - loss: 0.6774 - accuracy: 0.5812 - val_loss: 0.6711 - val_accuracy: 0.5667
Epoch 3/50
54/54 [==============================] - 1s 10ms/step - loss: 0.6613 - accuracy: 0.5971 - val_loss: 0.6654 - val_accuracy: 0.6033
Epoch 4/50
54/54 [==============================] - 1s 10ms/step - loss: 0.6513 - accuracy: 0.6035 - val_loss: 0.6664 - val_accuracy: 0.5533
Epoch 5/50
54/54 [==============================] - 1s 10ms/step - loss: 0.6301 - accuracy: 0.6435 - val_loss: 0.6717 - val_accuracy: 0.5700
Epoch 6/50
54/54 [==============================] - 0s 9ms/step - loss: 0.6145 - accuracy: 0.6688 - val_loss: 0.6508 - val_accuracy: 0.5933
Epoch 7/50
54/54 [==============================] - 1s 10ms/step - loss: 0.6031 - accuracy: 0.6735 - val_loss: 0.6670 - val_accuracy: 0.5933
Epoch 8/50
54/

For learning rate 0.00001, loss and accuracy is giving comparativley better result than learning rate 0, and learning rate 300. Here, in first epoch, loss is the most maximum value. Then loss decrease gradually all the way to epoch 50. And, accuracy in first epoch is lowest and it increases all the way to epoch 50.

***Question 4***

In [14]:
model_cnn = tf.keras.models.Sequential([
    tf.keras.layers.Conv2D(32, (3, 3), activation = 'relu', input_shape = (128, 128, 3)),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(64, (3, 3), activation = 'relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(64, (3, 3,), activation = 'relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(128, (3, 3), activation = 'relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(256, (3, 3), activation = 'relu'),
    tf.keras.layers.MaxPooling2D((2, 2,)),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(64, activation = 'relu'),
    tf.keras.layers.Dense(2, activation = 'softmax')
])


opt_cnn = tf.keras.optimizers.Adam(learning_rate = 0.00001)
model_cnn.compile (optimizer = opt_cnn, loss = 'sparse_categorical_crossentropy', metrics = ['accuracy'])
eval_cnn = model_cnn.fit(train_images, train_labels, epochs = 50, validation_split = 0.15)

Epoch 1/50
54/54 [==============================] - 7s 38ms/step - loss: 0.6937 - accuracy: 0.4924 - val_loss: 0.6917 - val_accuracy: 0.5667
Epoch 2/50
54/54 [==============================] - 1s 18ms/step - loss: 0.6909 - accuracy: 0.5541 - val_loss: 0.6907 - val_accuracy: 0.5767
Epoch 3/50
54/54 [==============================] - 1s 18ms/step - loss: 0.6893 - accuracy: 0.5800 - val_loss: 0.6891 - val_accuracy: 0.5700
Epoch 4/50
54/54 [==============================] - 1s 19ms/step - loss: 0.6873 - accuracy: 0.5865 - val_loss: 0.6885 - val_accuracy: 0.5700
Epoch 5/50
54/54 [==============================] - 1s 18ms/step - loss: 0.6856 - accuracy: 0.5900 - val_loss: 0.6881 - val_accuracy: 0.5700
Epoch 6/50
54/54 [==============================] - 1s 18ms/step - loss: 0.6836 - accuracy: 0.6212 - val_loss: 0.6867 - val_accuracy: 0.5800
Epoch 7/50
54/54 [==============================] - 1s 21ms/step - loss: 0.6811 - accuracy: 0.6182 - val_loss: 0.6848 - val_accuracy: 0.6000
Epoch 8/50
54

In [15]:
test_loss_fc, test_accuracy_fc = model_cnn.evaluate(test_images, test_labels)
print('Accuracy of the third fully connected network on the test dataset:', test_accuracy_fc)

32/32 [==============================] - 1s 16ms/step - loss: 0.6173 - accuracy: 0.6580
Accuracy of the third fully connected network on the test dataset: 0.6579999923706055


In [16]:
test_loss_cnn, test_accuracy_cnn = model_cnn.evaluate(test_images, test_labels)
print('Accuracy of the convolutional network on the test dataset:', test_accuracy_cnn)


32/32 [==============================] - 0s 8ms/step - loss: 0.6173 - accuracy: 0.6580
Accuracy of the convolutional network on the test dataset: 0.6579999923706055


***Question 5***

In [35]:
augmented_cnn_model = tf.keras.Sequential([
    tf.keras.layers.experimental.preprocessing.RandomFlip('horizontal_and_vertical', input_shape = (128, 128, 3)),
    tf.keras.layers.Conv2D(32, (3, 3), activation = 'relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(64, (3, 3), activation = 'relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D (64, (3, 3), activation = 'relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D (128, (3, 3), activation = 'relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D (256, (3, 3), activation = 'relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(64, activation = 'relu'),
    tf.keras.layers.Dense(2, activation = 'softmax')


])

augmented_cnn_model.compile(optimizer = tf.keras.optimizers.Adam(learning_rate = 0.00001),
                            loss = 'sparse_categorical_crossentropy',
                            metrics = ['accuracy'])
augmented_cnn_history = augmented_cnn_model.fit(train_images, train_labels, epochs = 50, validation_split = 0.15)

test_loss_augmented_cnn, test_accuracy_augmented_cnn = augmented_cnn_model.evaluate(test_images, test_labels)

print('Accuracy of augmented model score agains test data set:', test_accuracy_augmented_cnn)

Epoch 1/50
54/54 [==============================] - 4s 31ms/step - loss: 0.6928 - accuracy: 0.4941 - val_loss: 0.6928 - val_accuracy: 0.4867
Epoch 2/50
54/54 [==============================] - 1s 24ms/step - loss: 0.6917 - accuracy: 0.5035 - val_loss: 0.6924 - val_accuracy: 0.4867
Epoch 3/50
54/54 [==============================] - 1s 21ms/step - loss: 0.6908 - accuracy: 0.5029 - val_loss: 0.6920 - val_accuracy: 0.4867
Epoch 4/50
54/54 [==============================] - 1s 19ms/step - loss: 0.6900 - accuracy: 0.5024 - val_loss: 0.6914 - val_accuracy: 0.4900
Epoch 5/50
54/54 [==============================] - 1s 21ms/step - loss: 0.6886 - accuracy: 0.5200 - val_loss: 0.6912 - val_accuracy: 0.4867
Epoch 6/50
54/54 [==============================] - 1s 19ms/step - loss: 0.6872 - accuracy: 0.5082 - val_loss: 0.6909 - val_accuracy: 0.5000
Epoch 7/50
54/54 [==============================] - 1s 19ms/step - loss: 0.6860 - accuracy: 0.5329 - val_loss: 0.6903 - val_accuracy: 0.4900
Epoch 8/50
54

Here, maximum loss is 0.6928 in 1st epoch and minimum loss is 0.5965 in 50th epoch and maximum accuracy is 0.6888 in 50th epoch and minimum accuracy is 0.4941 in 1st epoch. The performance of augmented cnn model is not good as in th question 3 for learning rate 0.00001.